optuna/distributions.py

```diff
@@ -79,16 +78,37 @@ def _contains(self, param_value_in_internal_repr):
 
         raise NotImplementedError
 
-    @abc.abstractmethod
     def _asdict(self):
         # type: () -> Dict
 
-        raise NotImplementedError
+        return self.__dict__
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, type(self)):
+            return False
+
+        return self.__dict__ == other.__dict__
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        # type: () -> int
+
+        return hash(tuple(sorted(self.__dict__.items())))
+
+    def __repr__(self):
+        # type: () -> str
+
+        kwargs = ', '.join('{}={}'.format(k, v) for k, v in self.__dict__.items())
```

Comment:
Let's sort or somehow make the repr deterministic (until Python 3.6 where the `__dict__` order is guaranteed).

Reply:
I changed to apply `sorted` function (https://github.com/pfnet/optuna/pull/573/commits/01da38076bd4fef574ad8daa050174da2243255a). Thank you for pointing it out.